### ISSUE TITLE
Moving a Script Type from code file to another reverts entities attributes with that type to default values 

### ISSUE THREAD

**Issue by yaustar** [playcanvas/editor#396]
Steps to reproduce;

- Add a scripttype (testA.js) with an attribute
- Add the scripttype to an entity and change the default value
- Move code from testA.js to a new file testB.js
- Parse testA.js and testB.js
- Values on the entity are reverted to default

While generally not an issue, it can cause problems when refactoring, especially for webpack/typescript workflows where they may split the built script to several files to make parsing more manageable in the editor

**Comment by vkalpias** [comment]
I don't think this is a bug. Script attributes are associated with their asset. If you move code from one asset to another it's a different asset.

**Comment by Maksims** [comment]
> I don't think this is a bug. Script attributes are associated with their asset. If you move code from one asset to another it's a different asset.

Not really.
Attributes are stored on the entity, within script component data. And it only references it by the script name, not by an asset ID. So there probably should be no reason to lose that data. The only problem is when you parse testA.js, it detects no ScriptType, so the old script type is gone, so it might be removing related data on all entities where such script type was used.
Then when testB.js is parsed, it detects Script Type again and adds data back.

To solve it, it could simply avoid removing attributes from entity component data, when Script Type is removed. They will be updated when Script Type is present again.

**Comment by vkalpias** [comment]
Im not sure leaving attributes on an entity after they are gone from the asset is a good idea. Parsing two scripts are 2 different actions. Each asset stores the attribute definitions. If one asset loses the definitions leaving potential garbage on the entities is probably not worth this rare edge case.

**Comment by Maksims** [comment]
> Im not sure leaving attributes on an entity after they are gone from the asset is a good idea. Parsing two scripts are 2 different actions. Each asset stores the attribute definitions. If one asset loses the definitions leaving potential garbage on the entities is probably not worth this rare edge case.

Why leaving attributes on entities for yet not removed script (even not found one) would be bad?
If user removes script from the entity, then it and attributes will be removed.

There is no negatives on keeping attributes on entity script component, even if script type is not defined.

**Comment by vkalpias** [comment]
What if you remove attributes you no longer want? Won't the entity then store attributes that no longer exist forever and keep growing as you add / delete more?

**Comment by Maksims** [comment]
> What if you remove attributes you no longer want? Won't the entity then store attributes that no longer exist forever and keep growing as you add / delete more?

If you remove an attribute from the still defined Script Type, then it will remove such an attribute from the entity component. So keep this as is.
We talk here only specific case: when Script Type becomes undefined (removed entirely from the project), then do not clear attributes on entities.

When Script Type is parsed, it handles well when there are any variety of differences based on defined attributes on entity and Script Type, so no problem there too.

**Comment by vkalpias** [comment]
> We talk here only specific case: when Script Type becomes undefined (removed entirely from the project), then do not clear attributes on entities.

That might be OK I'll chat with the backend team
